### PR TITLE
rtio: Drop spsc usage

### PR DESCRIPTION
**Pull request by teburd**
No longer use the ring buffers as queues for the internal operations of rtio. The spsc (ring) queues may come back if io_uring's poll mode is implemented, where zero syscalls need to take place and would need to be reworked slightly.

* makes the multi-shot feature and cancellation simpler (see #56963)
* avoids head of line blocking as was a possibility in the previous iterations
* only one executor needed and much simplified
* improves performance of mpsc queues on non-smp systems (now faster than spsc)
* no measurable change in performance of rtio submit and completion pair (I measured it), small bump in ram usage
* bump in ram usage can be addressed now by adding a feature to share the sqe/cqe/mem block pools among rtio instances. There could be perhaps a system wide pool for each.

### CODE REVIEW

File: include/zephyr/rtio/rtio.h [hunk 1]
```diff
@@ -8,37 +8,33 @@
...
- * APIs.
+ * The completion of these operations typically provide one or more completion queue events.
+ *
+ * An executor takes the queues and determines how to perform each requested operation.
  */
 
 #ifndef ZEPHYR_INCLUDE_RTIO_RTIO_H_
 #define ZEPHYR_INCLUDE_RTIO_RTIO_H_
 
+#include "zephyr/sys/util.h"
```
Comment (yperess) [inline include/zephyr/rtio/rtio.h:31]: nit, use `<>`

File: include/zephyr/rtio/rtio.h [hunk 2]
```diff
@@ -387,11 +310,23 @@ struct rtio {
 	 */
 	atomic_t xcqcnt;
 
+	/* Submission queue object pool with free list */
+	struct rtio_iodev_sqe *sqe_pool;
+	struct rtio_mpsc sq_free;
+	uint16_t sqe_pool_sz;
```
Comment (yperess) [inline include/zephyr/rtio/rtio.h:316]: Should this be `const`?

File: include/zephyr/rtio/rtio.h [hunk 3]
```diff
@@ -387,11 +310,23 @@ struct rtio {
...
+	uint16_t sqe_pool_free;
+
+	/* Complete queue object pool with free list */
+	struct rtio_cqe *cqe_pool;
+	struct rtio_mpsc cq_free;
+	uint16_t cqe_pool_sz;
```
Comment (yperess) [inline include/zephyr/rtio/rtio.h:322]: Same here, `const`?

File: include/zephyr/rtio/rtio.h [hunk 4]
```diff
@@ -772,30 +730,66 @@ static inline uint32_t rtio_sqe_acquirable(struct rtio *r)
...
- *
- * @param r RTIO context
+ * @brief Acquire a complete queue event if available
  */
-static inline void rtio_sqe_drop_all(struct rtio *r)
+static inline struct rtio_cqe *rtio_cqe_acquire(struct rtio *r)
 {
-	rtio_spsc_drop_all(r->sq);
+	struct rtio_cqe *cqe =
+		CONTAINER_OF(rtio_mpsc_pop(&r->cq_free), struct rtio_cqe, q);
```
Comment (yperess) [inline include/zephyr/rtio/rtio.h:775]: Can you do this in one step? What happens if `rtio_mpsc_pop` returns NULL?

File: include/zephyr/rtio/rtio.h [hunk 5]
```diff
@@ -810,15 +804,26 @@ static inline void rtio_sqe_drop_all(struct rtio *r)
...
 #ifdef CONFIG_RTIO_CONSUME_SEM
 	if (k_sem_take(r->consume_sem, K_NO_WAIT) == 0) {
-		return rtio_spsc_consume(r->cq);
-	} else {
-		return NULL;
+		node = rtio_mpsc_pop(&r->cq);
+		if (node == NULL) {
+			return NULL;
+		}
+		cqe = CONTAINER_OF(node, struct rtio_cqe, q);
```
Comment (yperess) [inline include/zephyr/rtio/rtio.h:816]: Missing `r->cqe_pool_used--;`? Overall the two branches of the `ifdef` seem very similar, maybe we can have a
```
static inline int rtio_lock(struct rtio *r)
{
#ifdef CONFIG_RTIO_CONSUME_SEM
    return k_sem_tak(r->consume_sem, K_NO_WAIT);
#else
    return 0;
#endif
}
```

File: include/zephyr/rtio/rtio.h [hunk 6]
```diff
@@ -833,19 +838,24 @@ static inline struct rtio_cqe *rtio_cqe_consume(struct rtio *r)
...
-
+	node = rtio_mpsc_pop(&r->cq);
+	while (node == NULL) {
+		node = rtio_mpsc_pop(&r->cq);
+		Z_SPIN_DELAY(1);
 	}
+	cqe = CONTAINER_OF(node, struct rtio_cqe, q);
 #endif
 
 	return cqe;
```
Comment (yperess) [inline include/zephyr/rtio/rtio.h:861]: Is this missing `r->cqe_pool_used--;`?

File: include/zephyr/rtio/rtio.h [hunk 7]
```diff
@@ -311,29 +333,19 @@ struct rtio {
...
+	/* Mem block pool */
+	struct rtio_block_pool *block_pool;
 
 	/* Submission queue */
 	struct rtio_mpsc sq;
 
 	/* Completion queue */
 	struct rtio_mpsc cq;
-
-#ifdef CONFIG_RTIO_SYS_MEM_BLOCKS
```
Comment (yperess) [inline include/zephyr/rtio/rtio.h:396]: Are we removing the config?

**Review by yperess — DISMISSED** [review]

File: tests/subsys/rtio/rtio_api/src/test_rtio_api.c [hunk 8]
```diff
@@ -458,14 +424,49 @@ ZTEST(rtio_api, test_rtio_transaction)
...
+		cqe = rtio_cqe_consume(r);
+		rtio_cqe_release(r, cqe);
 	}
+
+	end_time = timing_counter_get();
+
+	uint64_t cycles = timing_cycles_get(&start_time, &end_time);
+	uint64_t ns = timing_cycles_to_ns(cycles);
+
+	zassert_true(ns < UINT64_C(1000000000), "Should take less than a second");
```
Comment (yperess) [inline tests/subsys/rtio/rtio_api/src/test_rtio_api.c:458]: Looks like this is failing

Comment (teburd) [reply tests/subsys/rtio/rtio_api/src/test_rtio_api.c:458]: Yeah right, on renode, which isn't easily ignored by the test case. Probably should remove the assert here? Up to you, I added them at your request :-)

Comment (yperess) [reply tests/subsys/rtio/rtio_api/src/test_rtio_api.c:458]: Feels odd to have a test with no asserts, I'll let you make the call on that. I do want to look at benchmarking tests but that's not slotted for me until 2024 :)

**Review by cfriedt — DISMISSED** [review]
LGTM

**Review by yperess — DISMISSED** [review]

**Comment by cfriedt** [comment]
@stephanosio - may need merge superpowers

**Review by stephanosio — CHANGES_REQUESTED** [review]
From the compliance check result:

>  SPACE_BEFORE_TAB: please, no space before tabs
File:include/zephyr/rtio/rtio.h
Line:708
 NEW_TYPEDEFS: do not add new typedefs
File:include/zephyr/rtio/rtio_mpsc.h
Line:41
 NEW_TYPEDEFS: do not add new typedefs
File:include/zephyr/rtio/rtio_mpsc.h
Line:49

No issue with adding new typedefs since they make sense in this case; but, the `SPACE_BEFORE_TAB` needs to be fixed.

**Review by yperess — APPROVED** [review]

**Review by stephanosio — APPROVED** [review]

**Comment by yperess** [comment]
@cfriedt PTAL when you get a chance. The only delta is the whitespace change.

**Review by cfriedt — APPROVED** [review]

**Comment by yperess** [comment]
@stephanosio can you use your super powers to merge this?

**Comment by stephanosio** [comment]
From compliance check workflow:

```
 NEW_TYPEDEFS: do not add new typedefs
File:include/zephyr/rtio/rtio_mpsc.h
Line:41
 NEW_TYPEDEFS: do not add new typedefs
File:include/zephyr/rtio/rtio_mpsc.h
Line:49
```

These new typedefs are justified. Merging.

**Pull request merged**